docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,966 @@
+        hook2->>capi: I can update []
+    capi->>apiserver: Create new MachineSet
+    loop For all Machines
+        capi->>apiserver: Replace machines
+    end
+```
+
+The user starts the process by creating a new VSphereMachineTemplate with the updated `template` value and updating the infrastructure template ref in the MachineDeployment:
+
+```diff
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: VSphereMachineTemplate
+metadata:
+-  name: md-1-2
++  name: md-1-3
+spec:
+  template:
+    spec:
+-     template: /Datacenter/vm/Templates/kubernetes-1-32-ubuntu
++     template: /Datacenter/vm/Templates/kubernetes-1-32-windows
+```
+
+```diff
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDeployment
+metadata:
+  name: m-cluster-vsphere-gaslor-md-0
+spec:
+  strategy:
+    type: InPlace
+    fallbackRollingUpdate:
+      maxUnavailable: 1
+  template:
+    spec:
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: VSphereMachineTemplate
+-       name: md-1-2
++       name: md-1-3
+```
+
+Both the `kcp-version-upgrade` and the `vsphere-vm-memory-update` extensions inform that they cannot handle any of the changes:
+
+```json
+{
+    "changes": ["infraMachine.spec.template"],
+}
+```
+
+```json
+{
+    "error": null,
+    "acceptedChanges": [],
+}
+```
+
+Since the fallback to machine replacement is a default strategy and always enabled, the MachineDeployment controller proceeds with the rollout process as it does today, replacing the old machines with new ones.
+
+### API Changes
+
+*All functionality related to In-Place Updates will be available only if the `InPlaceUpdates` feature flag is set to true.*
+
+#### External Update RuntimeExtension
+
+> TODO: we will add this later, after we get feedback from the first daft
+
+##### `CanUpdateMachine` endpoint
+##### Request
+> Requirements:
+> * Desired Machine/Bootstrap/InfraMachine changes
+
+##### Response
+> Requirements:
+> * Set of supported changes, probably an array of strings (the path in the object)
+> * Error
+
+##### `UpdateMachine` endpoint
+##### Request
+> Requirements:
+> *  Machine reference - namespace, name
+
+##### Response
+> Requirements:
+> * Result: [Success/Error/InProgress]
+> * Retry in X seconds
+
+### Security Model
+
+On the core CAPI side, the security model for this feature is very straightforward: CAPI controllers only require to read/create/update CAPI resources and those controllers are the only ones that need to modify the CAPI resources. Moreover, the controllers that need to perform these actions already have the necessary permissions over the resources they need to modify.
+
+However, each external updater should define their own security model. Depending on the mechanism used to update machines in-place, different privileges might be needed, from scheduling privileged pods to SSH access to the hosts. Moreover, external updaters  might need RBAC to read CAPI resources.
+
+### Risks and Mitigations
```

Comment:
I will ad short description about the risks due to the complexity of this change, which are mitigated by implementing this feature in incremental steps and by avoiding user facing changes in the first iteration.